### PR TITLE
Make bindings which produce UnsetValue revert to default value for property

### LINKED ISSUE #10110: DynamicResource is unsubscribing from the resource changes after it gets null/unset value once.
**Describe the bug**

In https://github.com/AvaloniaUI/Avalonia/pull/8166#issuecomment-1399403636 it was noticed that if some resources are defined only in one of many theme variants, DynamicResource stops working after switching between from variant with resource to the variant without the resource and back.
After some debugging, I found out that this issue is reproducible with plain DynamicResource without any theme variants as well.

**To Reproduce**

```csharp
[Fact]
public void Dynamically_Changing_Referenced_Resources_Works_With_DynamicResource()
{
    var xaml = @"
<UserControl xmlns='https://github.com/avaloniaui'
     xmlns:x='http://schemas.microsoft.com/winfx/2006/xaml'>
  <UserControl.Resources>
    <Color x:Key='color'>Red</Color>
    <SolidColorBrush x:Key='brush' Color='{DynamicResource color}' />
  </UserControl.Resources>
</UserControl>";

    var userControl = (UserControl)AvaloniaRuntimeXamlLoader.Load(xaml);

    Assert.Equal(Colors.Red, ((ISolidColorBrush)userControl.FindResource("brush")!).Color);

    userControl.Resources.Remove("color");
    Assert.Equal(default, ((ISolidColorBrush)userControl.FindResource("brush")!).Color);
    
    userControl.Resources.Add("color", Colors.Blue);
    Assert.Equal(Colors.Blue, ((ISolidColorBrush)userControl.FindResource("brush")!).Color);
}
```

**Additional context**

If resource wasn't found, DynamicResource returns Unset value, which is expected.
But BindingEntryBase when [receiving unset value clears current binding value.](https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Base/PropertyStore/BindingEntryBase.cs#L137-L142)
This code makes me think that it is an expected behavior. Still, I don't think that's how DynamicResource supposed to work.

Also, I tried to reproduce with simple subject and ToBinding set on the control, pushing unset value to the subject, but it was working perfectly.

**Desktop (please complete the following information):**

- Version: latest master

### PR DESCRIPTION
**Pull request by grokys**
## What does the pull request do?

As mentioned in #10110, #8600 introduced a problem whereby if a binding produces `UnsetValue` (or a binding error) then the binding is unsubscribed. 

This started happening because one of the aims of #8600 was to only have a subscription to one single active binding at a time [^1], but historically our behavior when a binding produces `UnsetValue` was for a binding or value with a lower priority to take effect; requiring multiple bindings to be subscribed.

As discussed in #10039, this historical behavior is different from WPF and I think can be changed. In WPF when a property produces `UnsetValue`, the property value is simply set to the default value of the property ignoring any lower priority values.

This PR makes that change. I'm leaving it as a draft for the moment to discuss the implications of this if any.

## Breaking changes

Changes the behavior of bindings. If anyone's binding was producing `UnsetValue` expecting the property value to revert to the value from a style etc then this will break.

## Fixed issues

Fixes #10110 

cc: @tomenscape for review/comments because this is something you've encountered.

[^1]: Previously all bindings were subscribed, even if their value was not visible which caused performance problems in some cases

### CODE REVIEW

File: src/Avalonia.Base/AvaloniaObject.cs [hunk 1]
```diff
@@ -270,8 +270,8 @@ public bool IsAnimating(AvaloniaProperty property)
         /// <param name="property">The property.</param>
         /// <returns>True if the property is set, otherwise false.</returns>
         /// <remarks>
-        /// Checks whether a value is assigned to the property, or that there is a binding to the
-        /// property that is producing a value other than <see cref="AvaloniaProperty.UnsetValue"/>.
+        /// Returns true if <paramref name="property"/> is a styled property which has a value
+        /// assigned to it or a binding targeting it; otherwise false.
```
Comment (grokys) [inline src/Avalonia.Base/AvaloniaObject.cs:274]: Due to the change in semantics, the meaning of `IsSet` had to change as well.

File: src/Avalonia.Base/PropertyStore/LocalValueBindingObserver.cs [hunk 2]
```diff
@@ -74,11 +78,21 @@ static void Execute(LocalValueBindingObserver<T> instance, BindingValue<T> value
                 LoggingUtils.LogIfNecessary(owner.Owner, property, value);
 
                 if (value.HasValue)
-                    owner.SetValue(property, value.Value, BindingPriority.LocalValue);
-                else if (value.Type != BindingValueType.DataValidationError)
-                    owner.ClearLocalValue(property);
+                {
+                    var effectiveValue = value.Value;
+                    if (property.ValidateValue?.Invoke(effectiveValue) == false)
```
Comment (grokys) [inline src/Avalonia.Base/PropertyStore/LocalValueBindingObserver.cs:83]: Fixes a bug here where we actually weren't validating the value in this case. Also added a unit test.

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0029508-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-all/nuget/v3/index.json) [PRBUILDID]

**Comment by tomenscape** [comment]
I've not run this code but it looks like exactly what I had in mind, so 👍 from me.

> This started happening because one of the aims of #8600 was to only have a subscription to one single active binding at a time, but historically our behavior when a binding produces UnsetValue was for a binding or value with a lower priority to take effect; requiring multiple bindings to be subscribed.

Not strictly related to this PR, but if Avalonia doesn't update the value of non-effective bindings, then what happens should they become effective later? Do we see a stale value, or is the current value retrieved from the source?

**Comment by grokys** [comment]
> Not strictly related to this PR, but if Avalonia doesn't update the value of non-effective bindings, then what happens should they become effective later? Do we see a stale value, or is the current value retrieved from the source?

Yes, when a binding becomes effective, it is subscribed, so the current value will be retrieved from the source.

**Comment by grokys** [comment]
I think this is ready for review now. The only thing I'm not sure about is the behavior of bindings _before_ they produce a value. Currently, there is an exception to the "one active binding subscription per property" rule, in that until a binding produces a value, multiple subscriptions can be active. This shouldn't be encountered in normal circumstances because standard bindings produce a value immediately in all cases except animation bindings.

I started making a change whereby bindings that don't produce a value when subscribed cause the default value to be produced, but some of the animation tests started failing. Problem is that there are so few animation tests that I'm unsure whether this will cause a real problem with animations, or if it's just how the tests were written.

Given the above, I think I'm going to leave this behavior as-is until it causes a problem, or we rewrite the animation stuff and add proper unit tests.

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0029796-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-all/nuget/v3/index.json) [PRBUILDID]

**Review by maxkatz6 — APPROVED** [review]

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0029870-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-all/nuget/v3/index.json) [PRBUILDID]

**Pull request merged**